storage/bootstrap/bootstrapper/fs/source.go

```diff
@@ -310,51 +389,123 @@ func (s *fileSystemSource) loadShardReadersDataIntoShardResult(
 	}
 	if seriesCachePolicy == series.CacheAllMetadata && shardRetriever == nil {
 		s.log.Errorf("shard retriever missing for shard: %d", shard)
-		s.handleErrorsAndUnfulfilled(resultLock, bootstrapResult, shard, tr, shardResult, timesWithErrors)
+		s.handleErrorsAndUnfulfilled(resultLock, runResult, shard, tr,
+			shardResult, timesWithErrors)
 		return
 	}
 
 	for _, r := range readers {
-		if shardResult == nil {
-			resultLock.RLock()
-			results := bootstrapResult.ShardResults()
-			var exists bool
-			shardResult, exists = results[shard]
-			resultLock.RUnlock()
-
-			if !exists {
-				resultLock.Lock()
+		var (
+			timeRange = r.Range()
+			start     = timeRange.Start
+			hasError  bool
+		)
+		switch run {
+		case bootstrapDataRunType:
+			if shardResult == nil {
+				resultLock.RLock()
+				results := runResult.data.ShardResults()
+				var exists bool
 				shardResult, exists = results[shard]
+				resultLock.RUnlock()
+
 				if !exists {
-					// NB(r): Wait until we have a reader to initialize the shard result
-					// to be able to somewhat estimate the size of it.
-					shardResult = result.NewShardResult(r.Entries(), bopts)
-					results[shard] = shardResult
+					resultLock.Lock()
+					shardResult, exists = results[shard]
+					if !exists {
+						// NB(r): Wait until we have a reader to initialize the shard result
+						// to be able to somewhat estimate the size of it.
+						shardResult = result.NewShardResult(r.Entries(), bopts)
+						results[shard] = shardResult
+					}
+					resultLock.Unlock()
 				}
-				resultLock.Unlock()
+
+				shardSeries = shardResult.AllSeries()
+			}
+		case bootstrapIndexRunType:
+			resultLock.Lock()
+			results := runResult.index.IndexResults()
+			indexBlockSegment, err = results.GetOrAddSegment(start,
+				ns.Options().IndexOptions(), bopts)
+			resultLock.Unlock()
+			if err != nil {
+				s.log.Errorf("unable to create index segment: %v", err)
+				hasError = true
 			}
+		default:
+			// Unreachable unless an internal method calls with a run type casted from int
+			panic(fmt.Errorf("invalid run type: %d", run))
 		}
 
-		var (
-			timeRange  = r.Range()
-			start      = timeRange.Start
-			hasError   = false
+		var numEntries int
+		if !hasError {
 			numEntries = r.Entries()
-		)
+		}
 		for i := 0; i < numEntries; i++ {
+			if run == bootstrapIndexRunType {
+				// If performing index run, then simply read the metadata and add to segment
+				id, tagsIter, _, _, err := r.ReadMetadata()
+				if err != nil {
+					s.log.Errorf("unable read metadata: %v", err)
+					hasError = true
+					continue
+				}
+
+				idBytes := id.Bytes()
+
+				resultLock.RLock()
+				exists, err := indexBlockSegment.ContainsID(idBytes)
+				resultLock.RUnlock()
+
+				if err == nil && !exists {
+					resultLock.Lock()
+					exists, err = indexBlockSegment.ContainsID(idBytes)
```

Comment:
could you move this to `IndexBootstrapResult.AddDoc(ident.ID, ident.Tags) error` instead? it'd make this code a lot cleaner